Move batch_norm to ATen/native, speed up

t-vi opened the pull request:
- Speed up the case of #12006 in the forward
- The backward still isn't as fast as one might hope (factor 2-3 in the #12006 case).
- More extensive benchmarking shows not so great performance compared
  to CuDNN for cases with many channels, e.g. bs=8-128 / c=1024 / f=1024.
- We change the meaning of save_mean and save_invstd (aka save_var) to accscalar to
  maintain reasonable precision.

Needless to say that I would happily separate the TensorAccessor fixes in a separate PR, as they're fixes and unrelated.

t-vi committed:
Move batch_norm to ATen/native, speed up

- Speed up the case of #12006 in the forward
- The backward still isn't as fast as one might hope (factor 2-3 in
the #12006 case.
- More extensive benchmarking shows not so great performance compared
  to CuDNN for cases with many channels, e.g.
  bs=8-128 / c=1024 / f=2014.
- We change the meaning of save_var to mean save_var for native batch
  norm. This appears to somewhat improve the numerical stability
  (i.e. makes TestNN.test_batch_norm_cudnn_half pass).

t-vi commented:
It's not the prettiest code, but I used this to arrive at the conclusion that for large numbers of features, it's still slow:
```
import torch
import timeit
import gc
import numpy

batch_sizes = [8,16,32,64,128]
channels    = [2,32,256,1024]
features    = [(16,),
               (32,),
               (64,),
               (128,),
               (256,),
               (1024,),
               (10240,),
               (102400,),
               (32,32),
               (64,64),
               (128,128),
               (32,32,32),
               (64,64,64)]

def run_bn(input, running_mean, running_var, weight, bias, training, backward):
    out = torch.nn.functional.batch_norm(input, running_mean, running_var, weight, bias, training, 0.1, 1e-5)
    if backward:
        grads = torch.autograd.grad(out, [input, weight, bias], torch.ones_like(out))
    torch.cuda.synchronize()

for bs in batch_sizes:
    for c in channels:
        for f in features:
            shape = (bs, c)+f
            size = numpy.prod(shape)
            if size < 100_000_000:
                running_mean = torch.randn(c, device='cuda')
                running_var = torch.randn(c, device='cuda').exp()
                weight = torch.randn(c, device='cuda', requires_grad=True)
                bias = torch.randn(c, device='cuda', requires_grad=True)
                input = torch.randn(shape, device='cuda', requires_grad=True)
                gc.collect()
                for training in [True, False]:
                    run_bn(input, running_mean, running_var, weight, bias, training, training)
                    torch.cuda.synchronize()
                    res1 = timeit.timeit('run_bn(input, running_mean, running_var, weight, bias, training, training)', number=100, globals=globals())
                    with torch.backends.cudnn.flags(enabled=False):
                      res2 = timeit.timeit('run_bn(input, running_mean, running_var, weight, bias, training, training)', number=100, globals=globals())
                    print ('{} | {} | {} | {} | {:.2f} | {:.2f} | {:.1f}'.format(bs, c, f, training, res1, res2, res1/res2))
```

t-vi commented:
Seems that the last rebase was less harmless than it looked...

ezyang commented:
I know what the problem is. You need to add `(aten, native_batch_norm)` to `aten/src/ATen/core/aten_intern
ed_strings.h` (and maybe others, but definitely this one).

CC @bwasti, this isn't going to be very good dev experience.

t-vi committed:
Add/remove interned strings for native/thnn batch norm and backward

Thank you, @ezyang for the hint! I was completely lost.

t-vi commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -23,6 +25,198 @@ namespace {
+  auto running_mean_a = conditional_accessor_1d<scalar_t>(running_mean);
+  auto running_var_a = conditional_accessor_1d<scalar_t>(running_var);
+
+  #pragma omp parallel for
```
Indeed. I fall for that. I'll use ATen's parallel_for instead.

I think one should also see whether some of the computation is better done using TensorIterators (currently I have very much kept the logic from TH(CU)NN except for splitting the cuda forward kernel).

t-vi committed:
use parallel_for, fix double backward's gradient mask parameter

- use parallel_for instead of using OMP myself.
  Thank you @vishwakftw for the review suggestion!
- The double backward used the backward's gradient mask instead
  of its own.

t-vi committed:
Change save_mean/var to accscalar, go back to thnn-like semantics

Thanks, ngimel for the advice.

t-vi commented:
I'm not entirely sure there isn't one, but I don't see the immediate connection to the failing TestScript.test_weak_script_function.

I think the correctness is there, though there are still performance improvements, in particular for the backward. What I'm not 100% certain about is whether removing thnn_batch_norm has compatibility implications.

SsnL commented:
Thanks a lot for doing this! I'm self-assigning so I remember to take a look. In the mean time, could you revert the submodule change?

t-vi committed:
Revert submodule change

Oops, thank you, Simon, for pointing this out.

t-vi committed:
Merge branch 'master' into native_batch_norm

SsnL commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -23,6 +25,195 @@ namespace {
+			       const Tensor& running_mean, const Tensor& running_var, bool train, double momentum, double eps) {
+
+  using accscalar_t = at::acc_type<scalar_t, false>;
+  Tensor output = at::native::empty_like(input);
```
Today I learned that you should always do `at::empty_like`. Yes it saves dispatch code, but it makes changing the dispatching implementation harder.

t-vi replied:
OK, I can adapt to that. :)

SsnL commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -23,6 +25,195 @@ namespace {
+  Tensor save_invstd;
+  const int64_t zero = 0;
+  if (train) {
+    save_mean = at::native::empty({n_input}, input.options());
```
Use `at::empty`.

ssnl reviewed (commented):
I know that I promised you that I'd take a look today, but it's really late and I just got to my room, and don't think that I can read kernels now. I will look tomorrow. In the meantime, if it's possible, could you briefly summarize what has changed between THNN version and the current version so that the kernels are faster? Thanks!

t-vi commented:
No worries, thanks for looking!
The kernel logic itself didn't change much, I mostly switched from DeviceTensor to PackedTensorAccessors. However, I changed the splitting of the kernel.
The kernel bits that changed:
- I amended the old "inference" kernel to the `batch_norm_transform_input_kernel` to deal both with
  running_mean/running_var (train=False) and save_mean/save_invstd (train=True).
  The former needs to calculate invstd from running_var, while the latter takes it directly.
  So now this is the only thing calculating out = w * (inp - mean) * invstd + b.
- I removed the bit done by `batch_norm_transform_input_kernel` from the old training forward kernel
  and renamed it to `batch_norm_collect_statistics_kernel`. It now only computes save_mean / save_var and  updates the running_mean / running_var as needed.
- This splitting of the training forward allows a favourable thread/block-parametrisation in the
  transform_input call, which used to be one main bottleneck in the forward.

`save_mean`/`save_std` have been changed to accscalar_t because we need the precision (previously, this was in a single kernel and also kept as accscalar_t and no-one cared about the poor backward). One might ask whether it would be desirable to (optionally) support accscalar_t weights for half (as cudnn does), but I didn't do this yet.

Finally, I also changed how the statistics gathering is parallelised by swapping dimensions 0 and 2 to have the larger as 2 in the calculation of the packed_accessor. This is because the parallelisation is done based on dimension 2 only.

t-vi commented:
Just a heads up: I've discussed kernels with the great folks working on apex sync batchnorm and implemented a backward kernel that is closer to what they use. If you wanted, I could drop in the updated backward kernel (which does similar changes as above, but in addition increases the accuracy of computation). Depending on your preference, I could add it in this PR, but by default, I'll keep it for a later PR (right now the backward kernel mostly the old THCUNN one).

t-vi committed:
Use at::... instead of at::native::...

Thank you, Simon, for the hint!

t-vi committed:
Merge branch 'master' into native_batch_norm

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -0,0 +1,444 @@
+  }
+
+  // sum over NumThreads within a warp
+  sum = warpSum(sum);
```
I'm a little bit lost here. I thought `warpSum` already sums all threads in a warp. Why do we need reduce again below?

t-vi replied:
This is unchanged logic from THCUNN, it took me a very close look to understand, too.

It's a two step warp sum to reduce from "one result per thread" to "<=32 results" to 1 result.
What the code does is that it has a first warpSum that reduces from whatever number of threads (<= 1024=32*32) to <= 32 intermediate sums. Then the "warp leader" will put that into shared memory. Then there is another round of reduction with warpSum  where the intermediate sums are reduced.
The alternative pattern is a reduction where you reduce down to <=32 first an then have a single round of reduction. (Which is done e.g. for the WeightNorm code.)
From what I've learned from @mcarilli and @jjsjann123, I think that it is a pattern that may offer some performance advantage, but not that much. In fact, my first stab at having a better backward uses the single reduction pattern.

SsnL replied:
This makes much more sense! Thanks for the explanation. Could you please update the function name and leave some comments here? Thanks! :)

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -0,0 +1,444 @@
+      strides[dim -1] = 1;
+    }
+  }
+  // evil trick to get adjusted 2d tensors to have large dimension last
```
So, IIUC, this is where you pick the larger of dim0 and dim2 to reduce over, right?

t-vi replied:
Indeed.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -0,0 +1,444 @@
+
+template <typename scalar_t, typename accscalar_t>
+__global__ void batch_norm_backward_kernel(
+    const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits> input,
```
Well I'm actually not sure if these should be `__restrict__` (other than `save_mean` and `save_invstd` and the output tensors). The other ones can be very well provided by the users.

t-vi replied:
My take on this is that the casual users will only provide input themselves, but I'm happy to remove them if you prefer.

SsnL replied:
Let's remove these traits. We shouldn't hurt out power users. :)

colesbury replied:
I don't understand. What's the argument for **not** making them `__restrict__`? Weren't they previously `__restrict__`?

t-vi replied:
I don't think they were `__restrict__` in the old code.
Personally, I don't see why people should set up overlapping tensors to trick themselves, but I can see how it's a bit safer when not assuming they don't.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -0,0 +1,444 @@
+  } else {
+    dim3 blocks_red(input.size(1));
+    dim3 threads_red(getNumThreads(input.size(2)));
+    batch_norm_collect_statistics_kernel<scalar_t, accscalar_t> <<<blocks_red, threads_red, 0, stream>>>
```
This is the part where stats collection and transform use different thread/block allocation, right?

t-vi replied:
Right.

ssnl reviewed (commented):
Kernels make a lot of sense! I'm a bit rusty on the warp reduction stuff so I didn't look at those code in details. I'm interested in seeing the benchmarking numbers. Could you post them?

Another thing that will speed this up a bit further is Welford algorithm, but it is out of scope of this PR.

SsnL commented:
Also, I assumed that CPU kernels didn't change much, right?

t-vi commented:
The CPU kernels are quite literally adapted from THNN.
I'll try to get a cleanish before/after benchmark (so far, I've mostly compared to cudnn).

t-vi committed:
Attempt to work around rocm not having std::sqrt

t-vi committed:
Merge branch 'master' into native_batch_norm

t-vi committed:
Remove __restrict__, comment on reduce, fix acctype tensors, remove unused params
- Remove __restrict__
  It seems difficult to support __restrict__ only on save_mean/save_var
  (and in particular not on running_mean/running_var). Unless we
  decide that supporting people doing funny stuff with them isn't
  worth it, we probably should jut not use restrict.
- Add some commentary on documentation.
- Fix input_options accscalaring to reflect that it's not inplace
  anymore.
- remove unused parameters from the statistics gathering kernel.

Thank you, Simon for the review comments, all errors are my own.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -78,36 +78,36 @@ struct Float2 {
 template <typename scalar_t, typename accscalar_t>
 struct SumOp {
-  __device__ SumOp(const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& t) : tensor(t) {}
+  __device__ SumOp(const PackedTensorAccessor<scalar_t, 3>& t) : tensor(t) {}
```
This might be a dumb question, considering my limited knowledge of `__restrict__`. Does `__restrict__` do anything here for this unary op?

t-vi replied:
This was just so I could pass around the tensoraccessor. :)

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -174,12 +190,12 @@ __device__ scalar_t reduce(Op op, PTA tensor, int plane) {
-    const PackedTensorAccessor<typename std::conditional<train, accscalar_t, scalar_t>::type, 1, at::RestrictPtrTraits> var_or_invstd,
-    const PackedTensorAccessor<scalar_t, 1, at::RestrictPtrTraits> weight,
-    const PackedTensorAccessor<scalar_t, 1, at::RestrictPtrTraits> bias,
+    const PackedTensorAccessor<scalar_t, 3> input,
```
I'm kinda curious how much does `__restrict__` gain us here.

SsnL replied:
If it's non-trivial, we might need to consider make it work for saved_mean and saved_invstd.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -365,7 +378,7 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_cuda_template(const Tensor& input_
   auto input_options = input_.options();
   if (input_options.dtype() == ScalarType::Half) {
-    input_options.dtype(ScalarType::Float);
+    input_options = input_options.dtype(ScalarType::Float);
```
:) thanks!

t-vi committed:
Only swap dimensions if feature dimension is really small

t-vi commented:
[There are a few cases that seem to get too slow](https://gist.github.com/t-vi/d72ef71a2184b5a9e3197f9a6763105d#gistcomment-2735273). :(

t-vi committed:
Consider number of planes in decision to swap dims

t-vi commented:
Is that the CI or my patch?

ezyang commented:
Yeah, aws dropped a new version of awscli which broke us. Rebase past #12787

t-vi committed:
Merge branch 'master' into native_batch_norm

t-vi committed:
Don't swap dimensions on backward

This leaves the backward alone, reducing the hazard of slowdowns
considerably.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -332,14 +335,16 @@ __global__ void batch_norm_backward_kernel(
+  if (in_dim == dim && (dim < 3 || (t.size(0) < t.size(2)) || (t.size(2) >= too_small_feature_set))) { // easy, if we don't need the evil trick
     return t.packed_accessor<scalar_t, dim>();
   }
 
```
What is this `AT_CHECK` on line 351? Can't we instead call `.contiguous()` rather than erroring out?

SsnL replied:
I would like to see some benchmark results on noncontig tensors, if this will be changed.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -332,14 +335,16 @@ __global__ void batch_norm_backward_kernel(
-static PackedTensorAccessor<scalar_t, dim> reshaped_packed_accessor(const Tensor& t) {
+static PackedTensorAccessor<scalar_t, dim> reshaped_packed_accessor(const Tensor& t, bool forward) {
+  constexpr int too_small_feature_set = 16;  // this is the maximum feature dimension when we swap
+  constexpr int few_planes = 256;
```
requesting more descriptive name / comment.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -387,21 +393,23 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_cuda_template(const Tensor& input_
-  dim3 blocks(batch_blocks, (input.size(1)+127)/128, feature_blocks);
-  dim3 threads(1, 128);
+  constexpr int max_blocks = 60000;
+  int feature_threads = std::min(getNumThreads(input.size(2)), 128);
```
Why is there a `min(*, 128)` here but not below on line 418 when you do the stats kernel?

SsnL replied:
I guess to better phrase the question: why does setting them this way benefit perf?

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -332,14 +335,16 @@ __global__ void batch_norm_backward_kernel(
 
 // TensorAccessor in which the last dimensions are collapsed or expanded as needed
```
This function currently does much more than just this comment. I hope that this can be extended to elaborate more on the design choices, i.e., to include when & why to swap dims.

t-vi commented:
So I scrapped a lot of what I had done because it was getting on the wrong path and re-thought some bits:
- In order to not reason too much about things and then changing performance characteristics, I think it is best to "welfordise" (compute mean and var in one go) the forward first, and indeed I have been very happy with that. (I'll push this after benchmarking.)
- I have also thought about `__restrict__` a bit more: The kernels follow a pattern of "do all computation first" and then "write outputs at the end". To me this would suggest that we can apply restrict without worries - the output, save_mean, save_var are our own anyway, but the running_var/mean are only written to at the end of the collect_stats kernel. I seem to see 5-10% speedup.
- For the input transform (eval / 2nd part in training) there is a balance to be struck between occupancy and memory. We want those for loops in the transform_input-kernel to loop some because it saves us reading the mean/var weight/bias all too often. This will again go have some heuristics for how many threads to use, but I think we'll be much happier when this is done.
- I replaced the funny TensorAccessor manipulation by a reshape and optional transpose.
- For the transpose (swap batch, feature), it seems that calling continuous is good but we have to balance the cost of copying with the advantage from swapping. There might be more room for experimentation, but for now I have some heuristic.

t-vi committed:
New strategy.

- Welfordization of forward statistics
- Move reshaping, swapping dims etc. out of accessor creation
- Try to have good occupancy in the transform input part

t-vi committed:
get rid of transposing in favour of adding batch threads if there are not many feature ones

t-vi committed:
test inlining tensoraccessor funtions

t-vi committed:
Merge branch 'master' into native_batch_norm

t-vi committed:
don't use inlining

t-vi committed:
fix 2d thread layout

t-vi committed:
finetune thread/block layout

t-vi commented:
I still don't know what makes the native batch norm administratively slower than THCUNNs. (There might be int vs. int64 for indexing, inlining of TensorAccessor things? There isn't really anything algorithmic in the evaluation forward...) :(
But the optimizations of the block/grid layout and Welford seem to compensate that, even if it would be preferable to have the benefits of both.

So I've heard that realistic channel feature combinations might be (64, 12544), (64, 3136), (256, 3136), (128, 3136), (128, 784), (512, 784),  (256, 784), (256, 196), (1024, 196), (512, 196), (512, 49), (2048, 49) and put that through my benchmark with several batch sizes.

From the things that take >1msec, these are the worst:

bs | channels | features | train | cudnn | native | slowness | reference | slowness | evaluation
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
128 | 2048 | 49 | e-fb | 1,4454 | 1,746 | 1,21 | 1,5894 | 1,1 |  
1024 | 512 | 49 | e-f | 0,8066 | 1,0233 | 1,27 | 0,9302 | 1,1 | bad
64 | 2048 | 49 | t-fb | 0,5744 | 1,1305 | 1,97 | 1,0425 | 1,08 |  
64 | 1024 | 196 | e-fb | 1,2717 | 1,2731 | 1 | 1,1907 | 1,07 |  
32 | 512 | 784 | e-fb | 1,1432 | 1,1382 | 1 | 1,0851 | 1,05 |  
64 | 512 | 784 | e-fb | 2,2656 | 2,2567 | 1 | 2,1661 | 1,04 |  
128 | 1024 | 196 | e-fb | 2,5212 | 2,5353 | 1,01 | 2,4556 | 1,03 |  
16 | 256 | 3136 | e-fb | 1,1148 | 1,1143 | 1 | 1,0993 | 1,01 |  
32 | 256 | 3136 | e-fb | 2,2072 | 2,2078 | 1 | 2,177 | 1,01 |  
64 | 256 | 3136 | e-fb | 4,4469 | 4,4497 | 1 | 4,4212 | 1,01 |  
64 | 256 | 784 | e-fb | 1,174 | 1,1753 | 1 | 1,1632 | 1,01 |  
128 | 512 | 196 | e-fb | 1,2994 | 1,308 | 1,01 | 1,2941 | 1,01 |  

Relative performance of fast kernels will be worse, e.g. bs=8,c=2048,f=49 takes 0.24msec vs. 0.066msec before.
And these are the best:

bs | channels | features | train | cudnn | native | slowness | reference | slowness | evaluation
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
64 | 128 | 3136 | t-f | 0,7592 | 1,0865 | 1,43 | 1,5893 | 0,68 | good
256 | 1024 | 196 | t-f | 1,6057 | 2,2576 | 1,41 | 3,3175 | 0,68 | good
256 | 2048 | 49 | t-f | 0,8236 | 1,587 | 1,93 | 2,369 | 0,67 | good
256 | 512 | 196 | t-f | 0,8657 | 1,1503 | 1,33 | 1,7539 | 0,66 | good
128 | 128 | 3136 | t-f | 1,5487 | 2,1475 | 1,39 | 3,2846 | 0,65 | good
256 | 256 | 784 | t-f | 1,5348 | 2,1127 | 1,38 | 3,2883 | 0,64 | good
256 | 128 | 784 | t-f | 0,8075 | 1,169 | 1,45 | 1,8861 | 0,62 | good
1024 | 512 | 49 | t-f | 1,3702 | 1,4977 | 1,09 | 2,46 | 0,61 | good
1024 | 256 | 196 | e-f | 1,2353 | 1,2558 | 1,02 | 2,4357 | 0,52 | good
1024 | 256 | 196 | t-f | 1,6817 | 2,3312 | 1,39 | 4,7348 | 0,49 | good

My other benchmark shows a smiliar thing.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -432,10 +466,15 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cuda_template(const Tenso
+  Tensor grad_input_cont;
   Tensor grad_weight_;
   Tensor grad_bias_;
+  auto input_cont = input_.reshape({input_.size(0), input_.size(1), -1});
```
you probably want to `.contiguous()` this before `reshape`. `reshape` doesn't guarantee contiguity.

however, I am a bit puzzled here. in theory, shouldn't `run kernel with noncontig data` always be at least as fast as `make contiguous + run kernel with contig data`?

t-vi replied:
I deliberately left it out, because I figured that I'd avoid the copy and the kernels don't need contiguity (except in the dimensions that get merged).
Regarding efficiency, I think it depends on whether it's possible to have a more efficient access pattern when making things contiguous than on non-contiguous data. In the end I figured that I might as well be going with non-contiguous data.
But I should have renamed the Tensors and will do now.

t-vi committed:
switch to 32 bit indexing

t-vi commented:
With 32 bit indexing, anything from the "real-world" benchmark taking more than 1msec seems to be not slower in native:

bs | channels | features | train | cudnn | native | slowness | reference | slowness | evaluation
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
32 | 512 | 784 | e-fb | 1,0846 | 1,0853 | 1 | 1,0851 | 1 |  
64 | 512 | 784 | e-fb | 2,1476 | 2,1521 | 1 | 2,1661 | 0,99 |  
16 | 64 | 12544 | e-fb | 1,0831 | 1,082 | 1 | 1,1056 | 0,98 |  
16 | 256 | 3136 | e-fb | 1,0748 | 1,0739 | 1 | 1,0993 | 0,98 |  
64 | 1024 | 196 | e-fb | 1,1715 | 1,1701 | 1 | 1,1907 | 0,98 |  
32 | 256 | 3136 | e-fb | 2,1151 | 2,1143 | 1 | 2,177 | 0,97 |  

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -405,12 +406,13 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_cuda_template(const Tensor& input_
-  auto bs = input_cont.size(0);
-  auto features = input_cont.size(2);
-  auto input = input_cont.packed_accessor<scalar_t, 3, RestrictPtrTraits>();
+  AT_CHECK(cuda::detail::canUse32BitIndexMath(input_reshaped), "Input is too large for batch_norm");
```
hmm can you instead template the whole thing and use `int32_t` if `canUse32BitIndexMath` and `int64_t` otherwise?

SsnL commented:
Very nice speed-up! :)

SsnL commented on aten/src/ATen/core/TensorAccessor.h:
```diff
@@ -91,56 +94,80 @@ class TensorAccessor<T,1,PtrTraits> : public TensorAccessorBase<T,1,PtrTraits> {
-  AT_HOST_DEVICE int64_t size(int64_t i) const { return sizes_[i]; }
+
+  // if index_t is not int64_t, we want to have an int64_t constructor
+  template <typename source_index_t, class = typename std::enable_if<std::is_same<source_index_t, int64_t>::value>::type>
```
I think this would be clearer if you get rid of the `typename source_index_t` arg, and only enable if `index_t != int64_t`.

t-vi replied:
I would need a hint how to do this. I tried all sorts of things and it didn't work out.
(But looking at the methods, I should probably use index_t more...)

SsnL replied:
Does this not work?
```
template<typename = std::enable_if<!std::is_same<index_t, int64_t>::value>::type>
```

t-vi replied:
No. I did try that. :) It gives:
`TensorAccessor.h(109): error: class "std::enable_if<false, void>" has no member "type"`
As far as I understand it, the reason is that you can have an "impossible template" (which is what enable_if produces in the false case) only as long as the template parameters are not fully given. So I need some unspecified bit, and after trying out various variants and trying to take something that looks similar to what `rgrep enable_if` gives, I settled on this. I don't like it, but it seems to not be too harmful.

SsnL replied:
I see. I've found that this works, but not sure if it is less ugly:
```
template<typename index_tt = index_t>
std::enable_if<!std::is_same<index_tt, int64_t>::value>::type
```

See https://godbolt.org/z/bIX4_O .

Keep it as-is also SGTM. BTW, if you want to use `enable_if_t`, we have it as `c10::guts::enable_if_t`.

t-vi replied:
I think that that introduces a return type for `f` which isn't applicable for the constructor.

t-vi committed:
templated switch between 64/32 bit indexing

t-vi committed:
fully specify template parameters for kernel calls for rocm

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -0,0 +1,536 @@
+							    const Tensor& save_mean, const Tensor& save_invstd, bool train, double epsilon, std::array<bool,3> grad_input_mask) {
+  return AT_DISPATCH_FLOATING_TYPES_AND_HALF(self.type(), "batch_norm_backward", [&] {
+      if (cuda::detail::canUse32BitIndexMath(self)) {
+	return batch_norm_backward_cuda_template<scalar_t, int32_t>(grad_out, self, weight, running_mean, running_var, save_mean, save_invstd, train, epsilon, grad_input_mask);
```
mis-indented

t-vi replied:
Ugh. I really need to get rid of all tabs. Thanks for catching this!

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -0,0 +1,536 @@
+						   const Tensor& running_mean, const Tensor& running_var, bool train, double momentum, double epsilon) {
+  return AT_DISPATCH_FLOATING_TYPES_AND_HALF(self.type(), "batch_norm", [&] {
+      if (cuda::detail::canUse32BitIndexMath(self)) {
+	return batch_norm_cuda_template<scalar_t, int32_t>(self, weight, bias, running_mean, running_var, train, momentum, epsilon);
```
mis-indented.

SsnL commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -0,0 +1,536 @@
+  // weight/bias) - which we only do once and have a for loop afterwards - with having many threads and blocks
+  // and good occupancy. Quiet likely, we could go with even more blocks than 1024.
+  // The various planes are independent, so we use blocks for them.
+  int tf = std::max<int>(getNumThreads(input.size(2)/4),
```
I wonder if these `/` in calculating `dim3`s should be ceil divisions instead.

t-vi replied:
No, for things passed into getNumThreads that's OK (because 0 will output some minimal number of threads) and some the others are actually intended as "b = CONST/a" with the guarantee that a*b<=CONST.

t-vi committed:
keep tabs on tabs or indenting will be ugly

t-vi committed:
save some contiguous calls
